File: runner/launcher/src/mill/launcher/MillProcessLauncher.java
package mill.launcher;

import static mill.constants.OutFiles.*;

import io.github.alexarchambault.nativeterm.NativeTerminal;
import io.github.alexarchambault.nativeterm.TerminalSize;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Stream;
import mill.client.ClientUtil;
import mill.constants.*;

public class MillProcessLauncher {

  static int launchMillNoDaemon(
      String[] args, OutFolderMode outMode, String[] runnerClasspath, String mainClass)
      throws Exception {
    final String sig = String.format("%08x", UUID.randomUUID().hashCode());
    final Path processDir =
        Paths.get(".").resolve(outFor(outMode)).resolve(millNoDaemon).resolve(sig);

    MillProcessLauncher.prepareMillRunFolder(processDir);

    final List<String> l = new ArrayList<>();
    l.addAll(millLaunchJvmCommand(outMode, runnerClasspath));
    Map<String, String> propsMap = ClientUtil.getUserSetProperties();
    for (String key : propsMap.keySet()) l.add("-D" + key + "=" + propsMap.get(key));
    l.add(mainClass);
    l.add(processDir.toAbsolutePath().toString());
    l.add(outMode.asString());
    l.addAll(millOpts(outMode));
    l.addAll(Arrays.asList(args));

    final ProcessBuilder builder = new ProcessBuilder().command(l).inheritIO();

    boolean interrupted = false;

    try {
      Process p = configureRunMillProcess(builder, processDir);
      return p.waitFor();

    } catch (InterruptedException e) {
      interrupted = true;
      throw e;
    } finally {
      if (!interrupted && Files.exists(processDir)) {
        // cleanup if process terminated for sure
        try (Stream<Path> stream = Files.walk(processDir)) {
          // depth-first
          stream.sorted(Comparator.reverseOrder()).forEach(p -> p.toFile().delete());
        }
      }
    }
  }

  static Process launchMillDaemon(Path daemonDir, OutFolderMode outMode, String[] runnerClasspath)
      throws Exception {
    List<String> l = new ArrayList<>(millLaunchJvmCommand(outMode, runnerClasspath));
    l.add("mill.daemon.MillDaemonMain");
    l.add(daemonDir.toFile().getCanonicalPath());
    l.add(outMode.asString());

    ProcessBuilder builder = new ProcessBuilder()
        .command(l)
        .redirectOutput(daemonDir.resolve(DaemonFiles.stdout).toFile())
        .redirectError(daemonDir.resolve(DaemonFiles.stderr).toFile());

    return configureRunMillProcess(builder, daemonDir);
  }

  static Process configureRunMillProcess(ProcessBuilder builder, Path daemonDir) throws Exception {

    Path sandbox = daemonDir.resolve(DaemonFiles.sandbox);
    Files.createDirectories(sandbox);

    builder.environment().put(EnvVars.MILL_WORKSPACE_ROOT, new File("").getCanonicalPath());
    if (System.getenv(EnvVars.MILL_EXECUTABLE_PATH) == null)
      builder.environment().put(EnvVars.MILL_EXECUTABLE_PATH, getExecutablePath());

    String jdkJavaOptions = System.getenv("JDK_JAVA_OPTIONS");
    if (jdkJavaOptions == null) jdkJavaOptions = "";
    String javaOpts = System.getenv("JAVA_OPTS");
    if (javaOpts == null) javaOpts = "";

    String opts = (jdkJavaOptions + " " + javaOpts).trim();
    if (!opts.isEmpty()) {
      builder.environment().put("JDK_JAVA_OPTIONS", opts);
    }

    builder.directory(sandbox.toFile());
    return builder.start();
  }

  static List<String> loadMillConfig(OutFolderMode outMode, String key) throws Exception {

    Path configFile = Paths.get("." + key);
    final Map<String, String> env = new HashMap<>();
    env.putAll(System.getenv());
    // Hardcode support for PWD because the graal native launcher has it set to the
    // working dir of the enclosing process, when we want it to be set to the working
    // dir of the current process
    final String workspaceDir = new java.io.File(".").getAbsoluteFile().getCanonicalPath();
    env.put("PWD", workspaceDir);
    env.put("WORKSPACE", workspaceDir);
    env.put("MILL_VERSION", mill.constants.BuildInfo.millVersion);
    env.put("MILL_BIN_PLATFORM", mill.constants.BuildInfo.millBinPlatform);

    if (Files.exists(configFile)) {
      return ClientUtil.readOptsFileLines(configFile.toAbsolutePath(), env);
    } else {
      for (String rootBuildFileName : CodeGenConstants.rootBuildFileNames) {
        Path buildFile = Paths.get(rootBuildFileName);
        if (Files.exists(buildFile)) {
          String[] config = cachedComputedValue(
              outMode,
              key,
              mill.constants.Util.readBuildHeader(
                  buildFile, buildFile.getFileName().toString()),
              () -> {
                Object conf = mill.launcher.ConfigReader.readYaml(
                    buildFile, buildFile.getFileName().toString());
                if (!(conf instanceof Map)) return new String[] {};
                @SuppressWarnings("unchecked")
                var conf2 = (Map<String, Object>) conf;

                if (!conf2.containsKey(key)) return new String[] {};
                if (conf2.get(key) instanceof List) {
                  @SuppressWarnings("unchecked")
                  var list = (List<String>) conf2.get(key);
                  String[] arr = new String[list.size()];
                  for (int i = 0; i < arr.length; i++) {
                    arr[i] = mill.constants.Util.interpolateEnvVars(list.get(i), env);
                  }
                  return arr;
                } else {
                  return new String[] {
                    mill.constants.Util.interpolateEnvVars(conf2.get(key).toString(), env)
                  };
                }
              });
          return Arrays.asList(config);
        }
      }
    }
    return List.of();
  }

  static List<String> millJvmOpts(OutFolderMode outMode) throws Exception {
    return loadMillConfig(outMode, "mill-jvm-opts");
  }

  static List<String> millOpts(OutFolderMode outMode) throws Exception {
    return loadMillConfig(outMode, "mill-opts");
  }

  static String millJvmVersion(OutFolderMode outMode) throws Exception {
    List<String> res = loadMillConfig(outMode, "mill-jvm-version");
    if (res.isEmpty()) return null;
    else return res.get(0);
  }

  static String millServerTimeout() {
    return System.getenv(EnvVars.MILL_SERVER_TIMEOUT_MILLIS);
  }

  static boolean isWin() {
    return System.getProperty("os.name", "").startsWith("Windows");
  }

  static String javaHome(OutFolderMode outMode) throws Exception {
    var jvmId = millJvmVersion(outMode);

    String javaHome = null;
    if (jvmId == null) {
      boolean systemJavaExists =
          new ProcessBuilder(isWin() ? "where" : "which", "java").start().waitFor() == 0;
      if (systemJavaExists && System.getenv("MILL_TEST_SUITE_IGNORE_SYSTEM_JAVA") == null) {
        jvmId = null;
      } else {
        jvmId = mill.client.BuildInfo.defaultJvmId;
      }
    }

    if (jvmId != null) {
      final String jvmIdFinal = jvmId;
      javaHome = cachedComputedValue0(
          outMode,
          "java-home",
          jvmId,
          () -> new String[] {CoursierClient.resolveJavaHome(jvmIdFinal).getAbsolutePath()},
          // Make sure we check to see if the saved java home exists before using
          // it, since it may have been since uninstalled, or the `out/` folder
          // may have been transferred to a different machine
          arr -> Files.exists(Paths.get(arr[0])))[0];
    }

    if (javaHome == null || javaHome.isEmpty()) javaHome = System.getProperty("java.home");
    if (javaHome == null || javaHome.isEmpty()) javaHome = System.getenv("JAVA_HOME");
    return javaHome;
  }

  static String javaExe(OutFolderMode outMode) throws Exception {
    String javaHome = javaHome(outMode);
    if (javaHome == null) return "java";
    else {
      final Path exePath = Paths.get(
          javaHome + File.separator + "bin" + File.separator + "java" + (isWin() ? ".exe" : ""));

      return exePath.toAbsolutePath().toString();
    }
  }

  static List<String> millLaunchJvmCommand(OutFolderMode outMode, String[] runnerClasspath)
      throws Exception {
    final List<String> vmOptions = new ArrayList<>();

    // Java executable
    vmOptions.add(javaExe(outMode));

    // sys props
    final Properties sysProps = System.getProperties();
    for (final String k : sysProps.stringPropertyNames()) {
      if (k.startsWith("MILL_")) {
        vmOptions.add("-D" + k + "=" + sysProps.getProperty(k));
      }
    }

    String serverTimeout = millServerTimeout();
    if (serverTimeout != null) vmOptions.add("-Dmill.server_timeout=" + serverTimeout);

    // extra opts
    vmOptions.addAll(millJvmOpts(outMode));

    vmOptions.add("-XX:+HeapDumpOnOutOfMemoryError");
    vmOptions.add("-cp");

    vmOptions.add(String.join(File.pathSeparator, runnerClasspath));

    return vmOptions;
  }

  static String[] cachedComputedValue(
      OutFolderMode outMode, String name, String key, Supplier<String[]> block) {
    return cachedComputedValue0(outMode, name, key, block, arr -> true);
  }

  static String[] cachedComputedValue0(
      OutFolderMode outMode,
      String name,
      String key,
      Supplier<String[]> block,
      Function<String[], Boolean> validate) {
    try {
      Path cacheFile = Paths.get(".").resolve(outFor(outMode)).resolve("mill-launcher/" + name);
      String[] value = null;
      if (Files.exists(cacheFile)) {
        String[] savedInfo = Files.readString(cacheFile).split("\n");
        if (savedInfo[0].equals(Escaping.literalize(key))) {
          value = Arrays.copyOfRange(savedInfo, 1, savedInfo.length);
          for (int i = 0; i < value.length; i++) value[i] = Escaping.unliteralize(value[i]);
        }
      }

      if (value != null && !validate.apply(value)) value = null;

      if (value == null) {
        value = block.get();
        String[] literalized = new String[value.length];
        for (int i = 0; i < literalized.length; i++) {
          literalized[i] = Escaping.literalize(value[i]);
        }

        Files.createDirectories(cacheFile.getParent());
        Files.write(
            cacheFile,
            (Escaping.literalize(key) + "\n" + String.join("\n", literalized)).getBytes());
      }
      return value;
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  static int getTerminalDim(String s, boolean inheritError) throws Exception {
    Process proc = new ProcessBuilder()
        .command("tput", s)
        .redirectOutput(ProcessBuilder.Redirect.PIPE)
        .redirectInput(ProcessBuilder.Redirect.INHERIT)
        // We cannot redirect error to PIPE, because `tput` needs at least one of the
        // outputstreams inherited so it can inspect the stream to get the console
        // dimensions. Instead, we check up-front that `tput cols` and `tput lines` do
        // not raise errors, and hope that means it continues to work going forward
        .redirectError(
            inheritError ? ProcessBuilder.Redirect.INHERIT : ProcessBuilder.Redirect.PIPE)
        .start();

    int exitCode = proc.waitFor();
    if (exitCode != 0) throw new Exception("tput failed");
    return Integer.parseInt(new String(proc.getInputStream().readAllBytes()).trim());
  }

  private static final AtomicReference<String> memoizedTerminalDims = new AtomicReference<>();

  private static final boolean canUseNativeTerminal;

  static {
    JLineNativeLoader.initJLineNative();

    boolean canUse;
    if (mill.constants.Util.hasConsole()) {
      try {
        NativeTerminal.getSize();
        canUse = true;
      } catch (Throwable ex) {
        canUse = false;
      }
    } else canUse = false;

    canUseNativeTerminal = canUse;
  }

  static void writeTerminalDims(boolean tputExists, Path daemonDir) throws Exception {
    String str;

    try {
      if (!mill.constants.Util.hasConsole()) str = "0 0";
      else {
        if (canUseNativeTerminal) {

          TerminalSize size = NativeTerminal.getSize();
          int width = size.getWidth();
          int height = size.getHeight();
          str = width + " " + height;
        } else if (!tputExists) {
          // Hardcoded size of a quarter screen terminal on 13" windows laptop
          str = "78 24";
        } else {
          str = getTerminalDim("cols", true) + " " + getTerminalDim("lines", true);
        }
      }
    } catch (Exception e) {
      str = "0 0";
    }

    // We memoize previously seen values to avoid causing lots
    // of upstream work if the value hasn't actually changed.
    // The upstream work could cause significant load, see
    //
    //    https://github.com/com-lihaoyi/mill/discussions/4092
    //
    // The cause is currently unknown, but this fixes the symptoms at least.
    //
    String oldValue = memoizedTerminalDims.getAndSet(str);
    if ((oldValue == null) || !oldValue.equals(str)) {
      Files.write(daemonDir.resolve(DaemonFiles.terminfo), str.getBytes());
    }
  }

  public static boolean checkTputExists() {
    try {
      getTerminalDim("cols", false);
      getTerminalDim("lines", false);
      return true;
    } catch (Exception e) {
      return false;
    }
  }

  public static void prepareMillRunFolder(Path daemonDir) throws Exception {
    // Clear out run-related files from the server folder to make sure we
    // never hit issues where we are reading the files from a previous run
    Files.deleteIfExists(daemonDir.resolve(DaemonFiles.terminfo));

    Path sandbox = daemonDir.resolve(DaemonFiles.sandbox);
    Files.createDirectories(sandbox);
    boolean tputExists = checkTputExists();

    writeTerminalDims(tputExists, daemonDir);
    Thread termInfoPropagatorThread = new Thread(
        () -> {
          try {
            while (true) {
              writeTerminalDims(tputExists, daemonDir);
              Thread.sleep(100);
            }
          } catch (Exception e) {
          }
        },
        "TermInfoPropagatorThread");
    termInfoPropagatorThread.setDaemon(true);
    termInfoPropagatorThread.start();
  }

  public static String getExecutablePath() {
    try {
      // Gets the location of the JAR or native image
      Path path = Paths.get(MillProcessLauncher.class
          .getProtectionDomain()
          .getCodeSource()
          .getLocation()
          .toURI());

      File file = path.toFile();
      return file.getAbsolutePath();
    } catch (java.net.URISyntaxException e) {
      throw new RuntimeException("Failed to determine Mill client executable path", e);
    }
  }
}

File: runner/launcher/src/mill/launcher/MillProcessLauncher.scala
package mill.launcher

import io.github.alexarchambault.nativeterm.NativeTerminal
import mill.api.internal.OneOrMore
import mill.client.ClientUtil
import mill.constants.*

import java.io.File
import java.util.UUID
import scala.jdk.CollectionConverters._

object MillProcessLauncher {

  def launchMillNoDaemon(
      args: Seq[String],
      outMode: OutFolderMode,
      runnerClasspath: Seq[os.Path],
      mainClass: String,
      useFileLocks: Boolean,
      workDir: os.Path,
      env: Map[String, String],
      millRepositories: Seq[String]
  ): Int = {
    val sig = f"${UUID.randomUUID().hashCode}%08x"
    val processDir =
      os.Path(OutFiles.OutFiles.outFor(outMode), workDir) / OutFiles.OutFiles.millNoDaemon / sig

    prepareMillRunFolder(processDir)

    val userPropsSeq = ClientUtil.getUserSetProperties().map { case (k, v) => s"-D$k=$v" }.toSeq

    val cmd = millLaunchJvmCommand(runnerClasspath, outMode, workDir, millRepositories) ++
      userPropsSeq ++
      Seq(mainClass, processDir.toString, outMode.asString, useFileLocks.toString) ++
      loadMillConfig(ConfigConstants.millOpts, workDir) ++
      args

    var interrupted = false
    val proc = configureRunMillProcess(cmd, processDir, workDir = workDir, env = env)
    try {
      proc.waitFor()
      proc.exitCode()
    } catch {
      case e: InterruptedException =>
        interrupted = true
        throw e
    } finally {
      if (!interrupted && os.exists(processDir)) os.remove.all(processDir)
    }
  }

  def launchMillDaemon(
      daemonDir: os.Path,
      outMode: OutFolderMode,
      runnerClasspath: Seq[os.Path],
      useFileLocks: Boolean,
      workDir: os.Path,
      env: Map[String, String],
      millRepositories: Seq[String]
  ): os.SubProcess = {
    val cmd = millLaunchJvmCommand(runnerClasspath, outMode, workDir, millRepositories) ++
      Seq("mill.daemon.MillDaemonMain", daemonDir.toString, outMode.asString, useFileLocks.toString)

    configureRunMillProcess(
      cmd,
      daemonDir,
      stdout = daemonDir / DaemonFiles.stdout,
      stderr = daemonDir / DaemonFiles.stderr,
      workDir = workDir,
      env = env
    )
  }

  private def configureRunMillProcess(
      cmd: Seq[String],
      daemonDir: os.Path,
      stdin: os.ProcessInput = os.Inherit,
      stdout: os.ProcessOutput = os.Inherit,
      stderr: os.ProcessOutput = os.Inherit,
      workDir: os.Path,
      env: Map[String, String]
  ): os.SubProcess = {
    val sandbox = daemonDir / DaemonFiles.sandbox
    os.makeDir.all(sandbox)

    val processEnv = Map(
      EnvVars.MILL_WORKSPACE_ROOT -> workDir.toString,
      EnvVars.MILL_ENABLE_STATIC_CHECKS -> "true"
    ) ++ (
      if (env.contains(EnvVars.MILL_EXECUTABLE_PATH)) Map.empty
      else Map(EnvVars.MILL_EXECUTABLE_PATH -> getExecutablePath)
    ) ++ {
      val jdkJavaOptions = env.getOrElse("JDK_JAVA_OPTIONS", "")
      val javaOpts = env.getOrElse("JAVA_OPTS", "")
      val opts = s"$jdkJavaOptions $javaOpts".trim
      if (opts.nonEmpty) Map("JDK_JAVA_OPTIONS" -> opts) else Map.empty
    }

    // destroyOnExit = false to prevent the daemon from being killed when the Mill client exits.
    // The daemon is a long-lived background process that should survive client disconnections.
    os.proc(cmd).spawn(
      cwd = sandbox,
      env = processEnv,
      stdin = stdin,
      stdout = stdout,
      stderr = stderr,
      destroyOnExit = false
    )
  }

  def loadMillConfig(key: String, workDir: os.Path = os.pwd): Seq[String] = {
    val configFile = workDir / s".$key"
    val env = mill.internal.Util.envForInterpolation(workDir)

    if (os.exists(configFile)) {
      ClientUtil.readOptsFileLines(configFile, env)
    } else {
      CodeGenConstants.rootBuildFileNames.asScala.toSeq
        .map(name => workDir / name)
        .find(os.exists(_))
        .map { buildFile =>
          val fileName = buildFile.last
          val headerData = Util.readBuildHeader(buildFile.toNIO, fileName)
          parseConfigFromHeader(headerData, key, env)
        }
        .getOrElse(Seq.empty)
    }
  }

  private def parseConfigFromHeader(
      headerData: String,
      key: String,
      env: Map[String, String]
  ): Seq[String] = {
    mill.internal.Util
      .parseBuildHeaderValue[OneOrMore[String]](headerData, key, default = OneOrMore(Nil))
      .value
      .map(item => Util.interpolateEnvVars(item, env.asJava))
  }

  def millServerTimeout: Option[String] =
    sys.env.get(EnvVars.MILL_SERVER_TIMEOUT_MILLIS)

  /**
   * Computes a fingerprint of JVM options that affect daemon behavior.
   * This includes options from .mill-jvm-opts file and JAVA_OPTS/JDK_JAVA_OPTIONS env vars.
   * When this fingerprint changes, the daemon should be restarted.
   */
  def computeJvmOpts(workDir: os.Path, env: Map[String, String]): Seq[String] = {
    val configOpts = loadMillConfig(ConfigConstants.millJvmOpts, workDir)
    val javaOpts = mill.api.internal.ParseArgs.parseShellArgs(env.getOrElse("JAVA_OPTS", ""))
    val jdkJavaOptions =
      mill.api.internal.ParseArgs.parseShellArgs(env.getOrElse("JDK_JAVA_OPTIONS", ""))
    configOpts ++ javaOpts ++ jdkJavaOptions
  }

  def isWin: Boolean = System.getProperty("os.name", "").startsWith("Windows")

  def javaHome(
      outMode: OutFolderMode,
      workDir: os.Path,
      millRepositories: Seq[String]
  ): Option[os.Path] = {
    val jvmVersion = loadMillConfig(ConfigConstants.millJvmVersion, workDir)
      .headOption
      .getOrElse(BuildInfo.defaultJvmVersion)

    val jvmIndexVersion = loadMillConfig(ConfigConstants.millJvmIndexVersion, workDir).headOption

    // Handle "system" specially - return None to use PATH-based Java lookup
    // (javaExe returns "java" when javaHome is None, using PATH lookup)
    if (jvmVersion == "system") None
    else Option.when(jvmVersion != null) {
      CoursierClient.resolveJavaHome(jvmVersion, jvmIndexVersion, outMode, millRepositories)
    }
  }

  def javaExe(
      outMode: OutFolderMode,
      workDir: os.Path,
      millRepositories: Seq[String]
  ): String = {
    javaHome(outMode, workDir, millRepositories) match {
      case None => "java"
      case Some(home) =>
        val exeName = if (isWin) "java.exe" else "java"
        (home / "bin" / exeName).toString
    }
  }

  def millLaunchJvmCommand(
      runnerClasspath: Seq[os.Path],
      outMode: OutFolderMode,
      workDir: os.Path,
      millRepositories: Seq[String]
  ): Seq[String] = {
    val millProps = sys.props.toSeq
      .filter(_._1.startsWith("MILL_"))
      .map { case (k, v) => s"-D$k=$v" }

    val serverTimeoutOpt = millServerTimeout.map(t => s"-Dmill.server_timeout=$t").toSeq

    // UTF-8 encoding options to fix Unicode character display issues on Windows
    val encodingOpts = Seq(
      "-Dfile.encoding=UTF-8",
      "-Dsun.stdout.encoding=UTF-8",
      "-Dsun.stderr.encoding=UTF-8"
    )

    Seq(javaExe(outMode, workDir, millRepositories)) ++
      millProps ++
      serverTimeoutOpt ++
      encodingOpts ++
      loadMillConfig(ConfigConstants.millJvmOpts, workDir) ++
      Seq("-cp", runnerClasspath.mkString(File.pathSeparator))
  }

  private def getTerminalDim(s: String, inheritError: Boolean): Int = {
    val result = os.proc("tput", s).call(
      stdin = os.Inherit,
      stdout = os.Pipe,
      stderr = if (inheritError) os.Inherit else os.Pipe,
      check = false
    )
    if (result.exitCode != 0) throw new Exception("tput failed")
    result.out.trim().toInt
  }

  // Returns native terminal size if available, avoiding calling getSize twice
  private def getNativeTerminalSize(): Option[(Int, Int)] = {
    JLineNativeLoader.initJLineNative()
    if (Util.hasConsole()) {
      try {
        val size = NativeTerminal.getSize
        Some((size.getWidth, size.getHeight))
      } catch {
        case _: Throwable => None
      }
    } else None
  }

  /** Get current terminal dimensions. Returns TerminalDimsResult with width/height. */
  def getTerminalDims(): DaemonRpc.TerminalDimsResult = {
    try {
      if (!Util.hasConsole()) DaemonRpc.TerminalDimsResult(None, None)
      else getNativeTerminalSize() match {
        case Some((width, height)) => DaemonRpc.TerminalDimsResult(Some(width), Some(height))
        case None if !tputExists => DaemonRpc.TerminalDimsResult(Some(78), Some(24))
        case None =>
          val width =
            try Some(getTerminalDim("cols", inheritError = true))
            catch { case _: Exception => None }
          val height =
            try Some(getTerminalDim("lines", inheritError = true))
            catch { case _: Exception => None }
          DaemonRpc.TerminalDimsResult(width, height)
      }
    } catch {
      case _: Exception => DaemonRpc.TerminalDimsResult(None, None)
    }
  }

  // Lazy so we only shell out to tput when canUseNativeTerminal is false
  private lazy val tputExists: Boolean = {
    try {
      getTerminalDim("cols", inheritError = false)
      getTerminalDim("lines", inheritError = false)
      true
    } catch {
      case _: Exception => false
    }
  }

  def prepareMillRunFolder(daemonDir: os.Path): Unit = {
    os.makeDir.all(daemonDir / DaemonFiles.sandbox)
  }

  def getExecutablePath: String = {
    try {
      os.Path(getClass.getProtectionDomain.getCodeSource.getLocation.toURI).toString
    } catch {
      case e: java.net.URISyntaxException =>
        throw new RuntimeException("Failed to determine Mill client executable path", e)
    }
  }
}
